File: app/src/main/java/com/software/jgodort/marvelheroes/network/MarvelSampleRepository.java
package com.software.jgodort.marvelheroes.network;

import com.software.jgodort.marvelheroes.network.callback.ResponseListener;

/**
 * Interface for MarvelSample Repository
 * Created by javie on 07/08/2017.
 */
public interface MarvelSampleRepository {

    void getSuperheroes(ResponseListener listener);
}

File: app/src/main/java/com/software/jgodort/marvelheroes/domain/repository/MarvelSampleRepository.kt
package com.software.jgodort.marvelheroes.domain.repository

import com.software.jgodort.marvelheroes.domain.MarvelRepository
import com.software.jgodort.marvelheroes.domain.SuperheroDomain
import com.software.jgodort.marvelheroes.network.MarvelAPI
import com.software.jgodort.marvelheroes.network.mapper.SuperheroRemoteMapper
import io.reactivex.Single
import javax.inject.Inject

class MarvelSampleRepository @Inject constructor(
  private val api: MarvelAPI,
  private val mapper: SuperheroRemoteMapper
) : MarvelRepository {

  override fun getSuperheroes(): Single<List<SuperheroDomain>> {

    return api.marvelSuperHeroes()
        .map { it.superheroes.map { mapper.mapFromRemote(it) } }
  }
}